Unique ID for every transfer (multiple token transfers per tx) and concurrent erc20 downloads

dshulyak opened the pull request:
closes: https://github.com/status-im/status-go/issues/1477

dshulyak committed:
Rewrite concurrent downloader and make erc20 downloader concurrent

dshulyak committed:
Generate unique ID for every transfer

ghost commented:
<!--prchecklist--> 
### Pull Request Checklist
- [ ] Have you updated the documentation, if impacted (e.g. [docs.status.im](https://docs.status.im/docs/build_status_go.html))?

status-im-auto commented:
### Jenkins Builds
<details>
<summary>Click to see older builds (15)</summary>

| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :heavy_check_mark: | dbb7499d | [#1](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-1482/1/) | 2019-06-06 07:49:08 | ~50 sec | `linux` | [:package: zip](https://status-go.ams3.digitaloceanspaces.com/status-go-desktop-190606-074818-dbb749.zip) |
| :heavy_check_mark: | dbb7499d | [#1](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-1482/1/) | 2019-06-06 07:51:54 | ~3 min | `ios` | [:package: zip](https://status-go.ams3.digitaloceanspaces.com/status-go-ios-190606-074817-dbb749.zip) |
| :heavy_check_mark: | dbb7499d | [#1](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-1482/1/) | 2019-06-06 07:54:52 | ~6 min | `android` | [:package: aar](https://status-go.ams3.digitaloceanspaces.com/status-go-android-190606-074815-dbb749.aar) |
| | | | | | | |
| :x: | d3ce2b78 | [#2](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-1482/2/) | 2019-06-07 08:50:33 | ~46 sec | `ios` | [:page_facing_up: log](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-1482/2/consoleText) |
| :heavy_check_mark: | d3ce2b78 | [#2](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-1482/2/) | 2019-06-07 08:50:37 | ~49 sec | `linux` | [:package: zip](https://status-go.ams3.digitaloceanspaces.com/status-go-desktop-190607-084947-d3ce2b.zip) |
| :x: | d3ce2b78 | [#2](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-1482/2/) | 2019-06-07 08:51:40 | ~1 min | `android` | [:page_facing_up: log](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-1482/2/consoleText) |
| | | | | | | |
| :heavy_check_mark: | f0b64094 | [#3](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-1482/3/) | 2019-06-07 08:55:40 | ~2 min | `linux` | [:package: zip](https://status-go.ams3.digitaloceanspaces.com/status-go-desktop-190607-085321-f0b640.zip) |
| :heavy_check_mark: | f0b64094 | [#3](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-1482/3/) | 2019-06-07 08:57:03 | ~3 min | `ios` | [:package: zip](https://status-go.ams3.digitaloceanspaces.com/status-go-ios-190607-085319-f0b640.zip) |
| :heavy_check_mark: | f0b64094 | [#3](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-1482/3/) | 2019-06-07 08:59:03 | ~5 min | `android` | [:package: aar](https://status-go.ams3.digitaloceanspaces.com/status-go-android-190607-085318-f0b640.aar) |
| | | | | | | |
| :heavy_check_mark: | a059cf8f | [#4](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-1482/4/) | 2019-06-07 09:20:06 | ~44 sec | `linux` | [:package: zip](https://status-go.ams3.digitaloceanspaces.com/status-go-desktop-190607-091921-a059cf.zip) |
| :heavy_check_mark: | a059cf8f | [#4](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-1482/4/) | 2019-06-07 09:23:46 | ~4 min | `ios` | [:package: zip](https://status-go.ams3.digitaloceanspaces.com/status-go-ios-190607-091920-a059cf.zip) |
| :heavy_check_mark: | a059cf8f | [#4](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-1482/4/) | 2019-06-07 09:24:59 | ~5 min | `android` | [:package: aar](https://status-go.ams3.digitaloceanspaces.com/status-go-android-190607-091919-a059cf.aar) |
| | | | | | | |
| :x: | bdc0a33c | [#5](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-1482/5/) | 2019-06-07 10:07:54 | ~40 sec | `ios` | [:page_facing_up: log](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-1482/5/consoleText) |
| :heavy_check_mark: | bdc0a33c | [#5](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-1482/5/) | 2019-06-07 10:08:02 | ~47 sec | `linux` | [:package: zip](https://status-go.ams3.digitaloceanspaces.com/status-go-desktop-190607-100714-bdc0a3.zip) |
| :x: | bdc0a33c | [#5](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-1482/5/) | 2019-06-07 10:08:49 | ~1 min | `android` | [:page_facing_up: log](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-1482/5/consoleText) |
</details>

| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :heavy_check_mark: | 402837a9 | [#6](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-1482/6/) | 2019-06-07 10:10:07 | ~31 sec | `linux` | [:package: zip](https://status-go.ams3.digitaloceanspaces.com/status-go-desktop-190607-100935-402837.zip) |
| :heavy_check_mark: | 402837a9 | [#6](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-1482/6/) | 2019-06-07 10:12:28 | ~2 min | `ios` | [:package: zip](https://status-go.ams3.digitaloceanspaces.com/status-go-ios-190607-100933-402837.zip) |
| :heavy_check_mark: | 402837a9 | [#6](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-1482/6/) | 2019-06-07 10:14:54 | ~5 min | `android` | [:package: aar](https://status-go.ams3.digitaloceanspaces.com/status-go-android-190607-100932-402837.aar) |
| | | | | | | |
| :x: | eaff713b | [#7](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-1482/7/) | 2019-06-13 06:26:25 | ~14 sec | `android` | [:page_facing_up: log](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-1482/7/consoleText) |
| :x: | eaff713b | [#7](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-1482/7/) | 2019-06-13 06:26:25 | ~11 sec | `linux` | [:page_facing_up: log](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-1482/7/consoleText) |
| :x: | eaff713b | [#7](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-1482/7/) | 2019-06-13 06:26:31 | ~19 sec | `ios` | [:page_facing_up: log](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-1482/7/consoleText) |

dshulyak committed:
Write better tests for reorg handling and improve handling description

dshulyak committed:
Renaming leftover

dshulyak committed:
Add logs and simplify lastKnownHeader

dshulyak committed:
Update accounts with all blocks in the set

dshulyak committed:
Encrypt database with same password taht is used for pfs db

dshulyak committed:
Rename ConcurrentRunner to AtomicGroup

dshulyak committed:
Add fast catchup after app was offline for longer than 15 blocks

When downloader is started it will go through following procedure:
- verify that the last synced block is still in canonical chain
- run fast indexer for every account from last synced block to head of the chain
- start watching block by block

dshulyak committed:
Remove ununes isSequence

dshulyak committed:
Increase timeout in test: bootstrap procedure may take longer now

dshulyak committed:
Remove unused code in database and adjust documentation
